### ISSUE TITLE
bugs in multinomial?

### ISSUE THREAD

**Issue by andresy** [torch/torch7#165]
Hi @nicholas-leonard 

Looking closely at multinomial, I see some strange behaviors...

- when an error occurs, it seems to be modifying the tensor of probabilities passed by the user:
> p = torch.rand(10)
> torch.multinomial(p, 3) -- ok
> torch.multinomial(p, 11) -- oops, error
> print(p:size())
  1
 10
[torch.LongStorage of size 2]

something wrong happened!

- i am not sure it is completely mathematically correct. Or is there something wrong in the random number generator? I am not sure. But if you try the following at home, with C=10,..,60 it works quite fine. With C = 80 or above, you see some weird spiky artefacts which should not exist. Any idea?

require 'torch'
require 'gnuplot'

torch.manualSeed(5555)

local C = 80
local N = 100000

-- some unbalanced data
local p = torch.Tensor(C)
for i=1,C do
   p[i] = i
end
local x = torch.multinomial(p, N, true)
gnuplot.axis{'','',0,''}
gnuplot.hist(x)

**Comment by nicholas-leonard** [comment]
@andresy I think there is a mistake concerning the default value of the replacement flag.
Try this:
```lua
torch.multinomial(p, 11, true)
 1  8  3  2  1  3  4  3  3  4  3
[torch.LongTensor of dimension 1x11]
```
So it's actually trying to sample without replace by default.

**Comment by andresy** [comment]
yes, i know, but what i mean if is that if an error occurs (like the one which occurs when forgetting the flag=true), then p is messed up. that should not happen.

[the second bug harder]

**Comment by nicholas-leonard** [comment]
This argcheck should have triggered : https://github.com/torch/torch7/blob/ef6a263b033831e0cd3435e6c412a809d0327c12/lib/TH/generic/THTensorRandom.c#L81

**Comment by andresy** [comment]
(in my example p should be still of size 10, not size 1x10)

**Comment by andresy** [comment]
yes, argcheck triggers, error occurs, but p is messed up. boo.

**Comment by nicholas-leonard** [comment]
Ah I see.

**Comment by nicholas-leonard** [comment]
I have to update TREPL to see it trigger...

**Comment by andresy** [comment]
:)

ok, for the second bug, i am not sure it is a bug in your code, or some subtility in the random generator.

**Comment by nicholas-leonard** [comment]
I am looking into that second part. Looks bad...

**Comment by nicholas-leonard** [comment]
@andresy Could it be a problem with `gnuplot.hist`?
```lua
require 'torch'
require 'gnuplot'
require 'torchx'
_ = require 'moses'

torch.manualSeed(5555)

local C = 80
local N = 1000000

-- some unbalanced data
local p = torch.range(1,C):double()
local x = torch.multinomial(p, N, true)

g = torch.group(x)
gc = _.map(g, function(k,v) return v.idx:size(1) end)
gnuplot.plot('Beuhp',p:type('torch.IntTensor'),torch.FloatTensor(gc),'|')
```
![beuhp](https://cloud.githubusercontent.com/assets/3520613/6543440/3f27684c-c4e9-11e4-8efb-21dfb17f6e80.png)
